syntax = "proto3";

package raft.proto;

service Raft {
    rpc AppendEntries(AppendEntriesArg) returns (AppendEntriesReply);
    rpc RequestVote(RequestVoteArg) returns (RequestVoteReply);
}

message LogEntry {
    uint64 term = 1;
    bytes command = 2;
}

message AppendEntriesArg {
    uint32 leaderId = 1;
    uint64 term = 2;
    uint64 prevLogIndex = 3;
    uint64 prevLogTerm = 4;
    uint64 leaderCommit = 5;
    repeated LogEntry entries = 6;
}

message AppendEntriesReply {
    bool success = 1;
    uint64 term = 2;
}

message RequestVoteArg {
    uint32 candidateId = 1;
    uint64 term = 2;
    uint64 lastLogIndex = 3;
    uint64 lastLogTerm = 4;
}

message RequestVoteReply {
    bool voteGranted = 1;
    uint64 term = 2;
}
